abd-clam/src/core/cluster/tree.rs

```diff
@@ -72,4 +78,372 @@ impl<I: Instance, U: Number, D: Dataset<I, U>> Tree<I, U, D> {
     pub const fn radius(&self) -> U {
         self.root.radius
     }
+
+    /// Saves a tree to a given location
+    ///
+    /// The path given will point to a newly created folder which will
+    /// store all necessary data for tree reconstruction.
+    ///
+    /// The directory structure looks like the following:
+    /// ```text
+    /// user_given_path/
+    ///      clusters/   <-- Clusters are serialized using their hex name.
+    ///      childinfo/  <-- Information about clusters immediate children.
+    ///      dataset/    <-- The serialized dataset.
+    ///      leaves.json <-- A json file of leaf names.
+    /// ````
+    ///
+    /// # Errors
+    /// Errors out on any directory or file creation issues
+    #[allow(clippy::missing_panics_doc)]
+    pub fn save(&self, path: &Path) -> Result<(), String> {
+        // Create our directory if needed
+        let dirbuilder = DirBuilder::new();
+
+        if !path.exists() {
+            dirbuilder.create(path).map_err(|e| e.to_string())?;
+        }
+
+        // Create cluster directory
+        let cluster_path = path.join("clusters");
+        dirbuilder.create(&cluster_path).map_err(|e| e.to_string())?;
+
+        // Create childinfo directory
+        let childinfo_path = path.join("childinfo");
+        dirbuilder.create(&childinfo_path).map_err(|e| e.to_string())?;
+
+        // Create dataset directory
+        let dataset_dir = path.join("dataset");
+        dirbuilder.create(&dataset_dir).map_err(|e| e.to_string())?;
+
+        // List of leaf clusters (Used for loading in the tree later)
+        let mut leaves: Vec<String> = vec![];
+
+        // Traverse the tree, serializing each cluster
+        let mut stack = vec![&self.root];
+        while let Some(cur) = stack.pop() {
+            let filename: String = cur.name();
+
+            // If the cluster is a parent, we push the children to the queue
+            if cur.is_leaf() {
+                leaves.push(filename.clone());
+            } else {
+                // Unwrapping is justified here because we validated that the cluster
+                // is a leaf before reaching this code
+                #[allow(clippy::unwrap_used)]
+                let [l, r] = cur.children().unwrap();
+                stack.push(l);
+                stack.push(r);
+            }
```

Comment:
Use a `match` statement on `cur.children()` instead of this. That will remove the use of `unwrap` altogether.